### PR TITLE
add `allow_fail` test attribute

### PR DESCRIPTION
**Pull request by pwoolcoc**
This change allows the user to add an `#[allow_fail]` attribute to
tests that will cause the test to compile & run, but if the test fails
it will not cause the entire test run to fail. The test output will
show the failure, but in yellow instead of red, and also indicate that
it was an allowed failure.

Here is an example of the output: http://imgur.com/a/wt7ga

### CODE REVIEW

**Comment by alexcrichton** [comment]
Thanks for the PR! We’ll periodically check in on it to make sure that @GuillaumeGomez or someone else from the team reviews it soon.

**Comment by Mark-Simulacrum** [comment]
Can you give some summary as to what the use cases for this would be? I can't think of anything where a failure would be okay (but not `should_fail`) right now.

**Comment by GuillaumeGomez** [comment]
I don't see the point of this add actually... :-/

**Comment by pwoolcoc** [comment]
This is basically for the same reasons you would set `allow_failures` in travis-ci. Maybe you want to commit a test for a feature or bug you haven't finished yet. You could set this instead of `ignore` so that the test always gets compiled, but won't fail the whole run if it foils at runtime. Or maybe it is an integration test that hits an outside API and you don't want the entire run to fail if there is an intermittent network issue that could cause the test to fail. The test is still shown in the output as failed, but it doesn't stop the whole run and the user doesn't have to do any `catch_unwind` tricks.

**Comment by Mark-Simulacrum** [comment]
Yeah, that helps understand the reasoning. I think I'm neutral on it then -- I don't think we really need it but would be fine having it.

**Comment by GuillaumeGomez** [comment]
I understand the reason but then: why not just use `no_run`? It'll compile and won't run, as you seem to need.

**Comment by pwoolcoc** [comment]
Well, `no_run` is just a rustdoc thing, it doesn't work for regular tests. And it is still helpful to know when a test succeeds, even if it's failure should be ignored.

**Comment by GuillaumeGomez** [comment]
Your change is included in rustdoc as well, but in there, it's pretty much useless. Then, for consistency, why not just add `no_run` into libsyntax instead of `allow_fail`?

**Comment by pwoolcoc** [comment]
I find `allow_fail` to be more useful because it has the same effect on the test suite as `no_run`, but still gives you information about the test.

**Comment by GuillaumeGomez** [comment]
That's just completely the opposite of the purpose of a test. I really don't agree with you on this functionality. However, maybe some other people might be interested so let's request opinions.

cc @rust-lang/compiler 
cc @rust-lang/core

**Comment by aturon** [comment]
@pwoolcoc I've wanted this functionality from time to time, and wouldn't be opposed to seeing it land. @alexcrichton, @brson, @nrc, I imagine you might have opinions also? I'm not really sure whose jurisdiction this falls under, tbh.

**Comment by alexcrichton** [comment]
This definitely falls under the category of "I wish we had custom test frameworks" to allow this level of customization, but I don't personally have many thoughts here. I think that the most appropriate owner "team-wise" nowadays is the dev-tools team, but in general libtest needs a lot of love not just in ad-hoc additions but also in overall direction.

I would not personally be opposed to landing this, although it would perhaps be nice to do it with a feature gate first.

**Comment by Mark-Simulacrum** [comment]
I think a feature gate is a good idea here, too.

**Comment by pnkfelix** [comment]
Am I right in inferring from the conversation that the reason that this `allow_fail` is different from `should_panic` is that an `allow_fail` test will always pass regardless of whether its execution invokes panic or not, while `should_panic` will not pass if its execution runs without invoking panic?

**Comment by Mark-Simulacrum** [comment]
Yes, that seems correct.

**Comment by pwoolcoc** [comment]
@pnkfelix you are mostly correct. a test marked `allow_fail` will still be reported as such in the test output, but it will _not_ cause the test run as a whole to fail.

**Comment by nikomatsakis** [comment]
There is also, of course, the `#[ignore]` flag which could be used for this; one problem I've found with *that* is that it is very easy to forget to "un-ignore" the tests. I imagine a similar thing could happen here, where you don't realize that the test is only passing because it is marked as `#[allow_fail]`.

Usually in this scenario what I do is to write the test so that it passes with the current behavior (e.g., by adding `#[should_panic]`, or by asserting that the "expected value" is in fact the current result, even though it's not what we eventually want) and leave some comments. This way, when I actually fix the code, I can't forget to update the test, since it starts to fail until I correct the behavior. 

I definitely think a feature-gate would be wise here, not sure how technically hard that would be to implement. I also think we might consider this as a "flag" to the existing `#[ignore]` attribute (e.g., `#[ignore(execute=true)]`) -- it feels conceptually similar, at least.

**Comment by steveklabnik** [comment]
> one problem I've found with that is that it is very easy to forget to "un-ignore" the tests. 

Some frameworks have a "pending" flag instead; what this does is, run the test, and on a failure, do nothing. On a success, it fails your build and says "hey you thought this was not a real test but it works"

**Comment by Mark-Simulacrum** [comment]
That seems very much like a "should-fail" flag; so in that respect we have a pending flag already.

**Comment by nagisa** [comment]
To me this sounds like a tag for spurious/wont-fix-just-yet failures. This is not comparable to `#[should_fail]` at all. `#[ignore]` seems like an alternative, but it really is not – it does not detect when the test begins passing again :)

My primary concern with this feature is that once you add this tag, it is very easy to forget to remove it later (similar applies to `#[ignore]`), but since this runs the test, it has an option to loudly remind you about unignoring this some how; not by eventually failing the test suite, though, as that would bring down the whole build altogether for no real reason.

Custom testing framework would be great.

**Comment by nikomatsakis** [comment]
> This is not comparable to `#[should_fail]` at all

I agree that the primary purpose of `#[should_panic]` is to write negative tests. But I think it can easily be used for the purpose of writing "wont-fix-just-yet" tests. In particular, the idea is to write a test that tests for the *current* behavior (even if is not what you want in the long term), along with a comment explaining what the *right* behavior would be -- if the current behavior is to panic, you might then use `#[should_panic]`. This has the added bonus that you find out if (a) the test starts passing or (b) if the test starts failing in some new, different way.

**Comment by nrc** [comment]
This seems a reasonable thing to add to me, but should definitely be behind a feature gate. I also agree that libtest needs some direction (and some RFC discussion), rather than just ad hoc extensions, but this seems like something that would be useful to experiment with first.

**Comment by nikomatsakis** [comment]
I agree a feature gate would make me feel better. I'm personally semi-opposed in that I think that it would be better practice to write the test so that it actively checks for the current (wrong) behavior, but I could imagine that people's workflows differ.

**Comment by pwoolcoc** [comment]
So, it looks like the consensus is "needs feature gate?" I will get that added and rebase, hopefully should be able to complete it tonight.

**Comment by carols10cents** [comment]
Since there's a feature gate now, and that's the only team-related feedback I see, I'm putting this back to waiting on review @GuillaumeGomez!

File: src/libsyntax/feature_gate.rs [hunk 1]
```diff
@@ -530,6 +530,7 @@ pub const BUILTIN_ATTRIBUTES: &'static [(&'static str, AttributeType, AttributeG
     ("derive", Normal, Ungated),
     ("should_panic", Normal, Ungated),
     ("ignore", Normal, Ungated),
+    ("allow_fail", Normal, Ungated),
```
Comment (GuillaumeGomez) [inline src/libsyntax/feature_gate.rs]: Isn't it supposed to be into an unstable gate first?

Comment (pwoolcoc) [inline src/libsyntax/feature_gate.rs]: Yes, I got sidetracked and haven't been able to go back and add the feature gate yet. I will push that change asap

**Comment by alexcrichton** [comment]
ping @pwoolcoc, just wanted to keep this on your radar!

**Comment by pwoolcoc** [comment]
@alexcrichton gah, sorry. I got distracted and haven't made it back to this yet. I will make sure this gets done at some point today.

**Comment by pwoolcoc** [comment]
Ok, I think this is in a good state.

r? @GuillaumeGomez

**Comment by GuillaumeGomez** [comment]
I just thought about something: should we add a warning saying that the `#[allow_fail]` test attribute should only be used on WIP code? That'd allow to make a difference between it and the other test attributes.

**Comment by pwoolcoc** [comment]
I'm not sure about that, only because not all the use cases for this are for WIP code. For example, I also use it for some integration tests that contact an external service that is not 100% available.

**Comment by GuillaumeGomez** [comment]
Can you at least display the result of the fail. For example:

```
some_test ... Success (result: Failure)
some_other_test ... Success (result: Success)
```

Or something along the line. Like that, even if it succeeds; you can get the underlying result.

**Comment by pwoolcoc** [comment]
It does show the test as "test allowed_to_fail ... FAILED (allowed)", but I'm certainly not married to that format: http://imgur.com/a/wt7ga

**Comment by GuillaumeGomez** [comment]
Oh my bad, for me it's good enough. As long as we have a notification, it's good for me. Then it's all good, thanks a lot @pwoolcoc!

@bors: r+

**Comment by jonhoo** [comment]
Does this have a issue tracking stabilization? /cc @GuillaumeGomez @arielb1

**Comment by GuillaumeGomez** [comment]
No. It should though. I'll open one.

**Pull request merged**